File: backend/API/serializers/utilisateur.py
from django.contrib.auth.hashers import make_password
from rest_framework import serializers
from ..models import Utilisateur
from . import BaseSerializer


class UtilisateurSerializer(BaseSerializer):
    """
    Sérializer du node Utilisateur

    Champs aditionnels :
        - pseudo : pseudo de l'utilisateur (obligatoire)
        - prenom : prénom de l'utilisateur (obligatoire)
        - nom : nom de l'utilisateur (obligatoire)
        - email : email de l'utilisateur (obligatoire)
        - password : mot de passe de l'utilisateur (obligatoire, write_only)
        Relations :
            - recherches_artistes : les artistes recherchés par l'utilisateur
            - regarder_interviews : les interviews regardées par l'utilisateur
            - regarder_extraits : les extraits regardés par l'utilisateur
            - recherches_questions : les questions recherchées par l'utilisateur
        Champs en read_only :
            - is_admin : indique si l'utilisateur est un administrateur
    """

    pseudo = serializers.CharField(required=True)
    prenom = serializers.CharField(required=True)
    nom = serializers.CharField(required=True)
    email = serializers.EmailField(required=True)
    password = serializers.CharField(write_only=True, required=True)
    is_admin = serializers.BooleanField(read_only=True)

    # Outputs
    recherches_artistes = serializers.SerializerMethodField(read_only=True)
    regarder_interviews = serializers.SerializerMethodField(read_only=True)
    regarder_extraits = serializers.SerializerMethodField(read_only=True)
    recherches_questions = serializers.SerializerMethodField(read_only=True)

    def __init__(self, *args, **kwargs):
        """
        node : Utilisateur
        """
        super().__init__(Utilisateur, *args, **kwargs)

    def get_recherches_artistes(self, utilisateur):
        """
        Renvoie un lien propre vers les artistes :
        """
        return self.get_url(
            "artiste-list", kwargs={"utilisateur_uuid": utilisateur.uuid}
        )

    def get_regarder_interviews(self, utilisateur):
        """
        Renvoie un lien propre vers les interviews :
        """
        return self.get_url(
            "interview-list", kwargs={"utilisateur_uuid": utilisateur.uuid}
        )

    def get_regarder_extraits(self, utilisateur):
        """
        Renvoie un lien propre vers les extraits :
        """
        return self.get_url(
            "extrait-list", kwargs={"utilisateur_uuid": utilisateur.uuid}
        )

    def get_recherches_questions(self, utilisateur):
        """
        Renvoie un lien propre vers les questions :
        """
        return self.get_url(
            "question-list", kwargs={"utilisateur_uuid": utilisateur.uuid}
        )

    def create(self, validated_data):
        """
        Création d'un utilisateur
        """
        # hash password
        validated_data["password"] = make_password(validated_data.pop("password"))
        return super().create(validated_data)

    def update(self, utilisateur, validated_data):
        """
        Modification d'un utilisateur
        """
        # hash password
        pwd = validated_data.pop("password", None)
        if pwd:
            validated_data["password"] = make_password(pwd)
        return super().update(utilisateur, validated_data)

File: frontend/src/model/utilisateur.js
import Artiste from "@model/artiste.js";
import Extrait from "@model/extrait.js";
import Interview from "@model/interview.js";
import Model from "@model/model.js";
import Question from "@model/question.js";

export default class Utilisateur extends Model {
    #pseudo;
    #prenom;
    #nom;
    #email;
    #password;
    #is_admin;
    #recherches_artistes;
    #regarder_interviews;
    #regarder_extraits;
    #recherches_questions;

    constructor({ uuid, pseudo, prenom, nom, email, is_admin, recherches_artistes, regarder_interviews, regarder_extraits, recherches_questions }) {
        super(uuid);
        this.#pseudo = pseudo;
        this.#prenom = prenom;
        this.#nom = nom;
        this.#email = email;
        this.#is_admin = is_admin;
        this.#recherches_artistes = recherches_artistes;
        this.#regarder_interviews = regarder_interviews; // liste de prommesses d'interviews regarder par le user
        this.#regarder_extraits = regarder_extraits; // liste de prommesses d'extraits regarder par le user
        this.#recherches_questions = recherches_questions;
    }

    static get endpoint() { return "utilisateurs"; }

    get pseudo() { return this.#pseudo; }
    set pseudo(value) { this.#pseudo = this.validateString(value, "pseudo"); }

    get prenom() { return this.#prenom; }
    set prenom(value) { this.#prenom = this.validateString(value, "prenom"); }

    get nom() { return this.#nom; }
    set nom(value) { this.#nom = this.validateString(value, "nom"); }

    get email() { return this.#email; }
    set email(value) { this.#email = this.validateString(value, "email"); }

    get password() { return this.#password; }
    set password(value) { this.#password = this.validateString(value, "password"); } // hash à gérer côté backend

    get is_admin() { return this.#is_admin; }
    set is_admin(value) { this.#is_admin = !!value; }

    async recherches_artistes(args) { return await this.fetchList(this.#recherches_artistes, Artiste, args); }

    async regarder_interviews(args) { return await this.fetchList(this.#regarder_interviews, Interview, args); }

    async regarder_extraits(args) { return await this.fetchList(this.#regarder_extraits, Extrait, args); }

    async recherches_questions(args) { return await this.fetchList(this.#recherches_questions, Question, args); }

    /**
     * Connecte un utilisateur à un artiste
     * @param {Artiste} artiste 
     */
    connect_artiste(artiste) {
        this.connect(this.#recherches_artistes, { 'uuid': artiste.uuid });
    }

    /**
     * Déconnecte un utilisateur d'un artiste
     * @param {Artiste} artiste 
     */
    disconnect_artiste(artiste) {
        this.disconnect(this.#recherches_artistes, artiste);
    }

    /**
     * Connecte un utilisateur à une interview
     * @param {Interview} interview 
     */
    connect_interview(interview) {
        this.connect(this.#regarder_interviews, { 'uuid': interview.uuid });
    }

    /**
     * Déconnecte un utilisateur d'une interview
     * @param {Interview} interview 
     */
    disconnect_interview(interview) {
        this.disconnect(this.#regarder_interviews, interview);
    }

    /**
     * Connecte un utilisateur à un extrait
     * @param {Extrait} extrait 
     */
    connect_extrait(extrait) {
        this.connect(this.#regarder_extraits, { 'uuid': extrait.uuid });
    }

    /**
     * Déconnecte un utilisateur d'un extrait
     * @param {Extrait} extrait 
     */
    disconnect_artiste(extrait) {
        this.disconnect(this.#regarder_extraits, extrait);
    }

    /**
     * Connecte un utilisateur à une question
     * @param {Question} question 
     */
    connect_artiste(question) {
        this.connect(this.#recherches_questions, { 'uuid': question.uuid });
    }

    /**
     * Déconnecte un utilisateur d'une question
     * @param {Question} question 
     */
    disconnect_artiste(question) {
        this.disconnect(this.#recherches_questions, question);
    }

    fromJSON(json) {
        super.fromJSON(json);
        this.#pseudo = json.pseudo;
        this.#prenom = json.prenom;
        this.#nom = json.nom;
        this.#email = json.email;
        this.#is_admin = json.is_admin;
        this.#recherches_artistes = json.recherches_artistes;
        this.#regarder_interviews = json.regarder_interviews;
        this.#regarder_extraits = json.regarder_extraits;
        this.#recherches_questions = json.recherches_questions;
        return this;
    }

    toJSON(json = {}) {
        json = super.toJSON(json)
        if (this.name) json['name'] = this.name;
        if (this.pseudo) json['pseudo'] = this.pseudo;
        if (this.prenom) json['prenom'] = this.prenom;
        if (this.nom) json['nom'] = this.nom;
        if (this.email) json['email'] = this.email;
        if (this.password) json['password'] = this.password;
        if (this.is_admin) json['is_admin'] = this.is_admin;
        return json;
    }
}